File: SWEagent/tests/test-parsing.test.ts
/**
 * Parsing tests converted from test_parsing.py
 */

import { describe, it, expect } from '@jest/globals';
import {
  ActionOnlyParser as ActionParser,
  EditFormatParser as EditFormat,
  FunctionCallingParser,
  Identity,
  JsonParser,
  ThoughtActionParser,
  XMLThoughtActionParser,
} from '../src/agent/tools/parsing';
import { Command } from '../src/agent/types';
import { FormatError, FunctionCallingFormatError } from '../src/exceptions';

describe('Parsing', () => {
  describe('ActionParser', () => {
    it('should parse valid commands', () => {
      const parser = new ActionParser();
      const command: Command = {
        name: 'ls',
      };

      const [thought, action] = parser.parse({ message: 'ls -l' }, [command]);

      expect(thought).toBe('');
      expect(action).toBe('ls -l');
    });

    it('should throw FormatError for invalid commands', () => {
      const parser = new ActionParser();
      const command: Command = {
        name: 'ls',
      };

      expect(() => {
        parser.parse({ message: 'invalid command' }, [command]);
      }).toThrow(FormatError);
    });
  });

  describe('ThoughtActionParser', () => {
    it('should parse thought and action with code blocks', () => {
      const parser = new ThoughtActionParser();
      const modelResponse = "Let's look at the files in the current directory.\n```\nls -l\n```";

      const [thought, action] = parser.parse({ message: modelResponse }, []);

      expect(thought).toBe("Let's look at the files in the current directory.\n");
      expect(action).toBe('ls -l\n');
    });

    it('should throw FormatError when no code block', () => {
      const parser = new ThoughtActionParser();

      expect(() => {
        parser.parse({ message: 'No code block' }, []);
      }).toThrow(FormatError);
    });
  });

  describe('XMLThoughtActionParser', () => {
    it('should parse XML command tags', () => {
      const parser = new XMLThoughtActionParser();
      const modelResponse = "Let's look at the files in the current directory.\n<command>\nls -l\n</command>";

      const [thought, action] = parser.parse({ message: modelResponse }, []);

      expect(thought).toBe("Let's look at the files in the current directory.");
      expect(action).toBe('ls -l');
    });

    it('should throw FormatError when no command tags', () => {
      const parser = new XMLThoughtActionParser();

      expect(() => {
        parser.parse({ message: 'No command tags' }, []);
      }).toThrow(FormatError);
    });
  });

  describe('EditFormat', () => {
    it('should parse edit format with code blocks', () => {
      const parser = new EditFormat();
      const modelResponse = "Let's replace the contents.\n```\nimport os\nos.listdir()\n```";

      const [thought, action] = parser.parse({ message: modelResponse }, []);

      expect(thought).toBe("Let's replace the contents.\n");
      expect(action).toBe('import os\nos.listdir()\n');
    });

    it('should throw FormatError when no code block', () => {
      const parser = new EditFormat();

      expect(() => {
        parser.parse({ message: 'No code block' }, []);
      }).toThrow(FormatError);
    });
  });

  describe('Identity', () => {
    it('should return input as is', () => {
      const parser = new Identity();
      const modelResponse = 'Return as is';

      const [thought, action] = parser.parse({ message: modelResponse }, []);

      expect(thought).toBe(modelResponse);
      expect(action).toBe(modelResponse);
    });
  });

  describe('JsonParser', () => {
    it('should parse valid JSON', () => {
      const parser = new JsonParser();
      const modelResponse = '{"thought": "List files", "command": {"name": "ls", "arguments": {"path": "."}}}';

      const [thought, action] = parser.parse({ message: modelResponse }, []);

      expect(thought).toBe('List files');
      expect(action).toBe('ls .');
    });

    it('should throw FormatError for invalid JSON', () => {
      const parser = new JsonParser();

      expect(() => {
        parser.parse({ message: 'Not a JSON' }, []);
      }).toThrow(FormatError);
    });

    it('should throw FormatError for missing keys', () => {
      const parser = new JsonParser();
      const missingKeys = '{"thought": "Missing command key"}';

      expect(() => {
        parser.parse({ message: missingKeys }, []);
      }).toThrow(FormatError);
    });
  });

  describe('FunctionCallingParser', () => {
    it('should parse function calls successfully', () => {
      const parser = new FunctionCallingParser();
      const command: Command = {
        name: 'ls',
      };

      const modelResponse = {
        message: "Let's list the files",
        tool_calls: [
          {
            function: {
              name: 'ls',
              arguments: '{}',
            },
          },
        ],
      };

      const [thought, action] = parser.parse(modelResponse, [command]);

      expect(thought).toBe("Let's list the files");
      expect(action).toBe('ls');
    });

    it('should throw error for missing tool_calls', () => {
      const parser = new FunctionCallingParser();
      const command: Command = {
        name: 'ls',
      };

      expect(() => {
        parser.parse({ message: 'No tool calls' }, [command]);
      }).toThrow(FormatError);
    });

    it('should throw error for multiple tool calls', () => {
      const parser = new FunctionCallingParser();
      const command: Command = {
        name: 'ls',
      };

      const multipleCalls = {
        message: 'Multiple calls',
        tool_calls: [{ function: { name: 'ls', arguments: '{}' } }, { function: { name: 'cd', arguments: '{}' } }],
      };

      expect(() => {
        parser.parse(multipleCalls, [command]);
      }).toThrow(FormatError);
    });

    it('should throw error for invalid command', () => {
      const parser = new FunctionCallingParser();
      const command: Command = {
        name: 'ls',
      };

      const invalidCommand = {
        message: 'Invalid command',
        tool_calls: [
          {
            function: {
              name: 'invalid',
              arguments: '{}',
            },
          },
        ],
      };

      expect(() => {
        parser.parse(invalidCommand, [command]);
      }).toThrow(FormatError);
    });

    it('should throw error for invalid JSON arguments', () => {
      const parser = new FunctionCallingParser();
      const command: Command = {
        name: 'ls',
      };

      const invalidJson = {
        message: 'Invalid JSON',
        tool_calls: [
          {
            function: {
              name: 'ls',
              arguments: 'invalid json',
            },
          },
        ],
      };

      expect(() => {
        parser.parse(invalidJson, [command]);
      }).toThrow(FormatError);
    });
  });

  describe('FunctionCallingParser error messages', () => {
    it('should format error message correctly', () => {
      const parser = new FunctionCallingParser();
      const error = new FunctionCallingFormatError('test', 'missing');
      const message = parser.formatErrorMessage(error);

      expect(message).toContain('did not use any tool calls');
    });
  });
});

File: cpp_generated/test-parsing.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/SWEagent/tests/test-parsing.test.h"

void Main(void)
{
    describe(std::string("Parsing"), [=]() mutable
    {
        describe(std::string("ActionParser"), [=]() mutable
        {
            it(std::string("should parse valid commands"), [=]() mutable
            {
                auto parser = std::make_shared<ActionParser>();
                auto command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                auto [thought, action] = parser->parse(object{
                    object::pair{std::string("message"), std::string("ls -l")}
                }, array<std::shared_ptr<Command>>{ command });
                expect(thought)->toBe(string_empty);
                expect(action)->toBe(std::string("ls -l"));
            }
            );
            it(std::string("should throw FormatError for invalid commands"), [=]() mutable
            {
                shared parser = std::make_shared<ActionParser>();
                shared command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), std::string("invalid command")}
                    }, array<std::shared_ptr<Command>>{ command });
                }
                )->toThrow(FormatError);
            }
            );
        }
        );
        describe(std::string("ThoughtActionParser"), [=]() mutable
        {
            it(std::string("should parse thought and action with code blocks"), [=]() mutable
            {
                auto parser = std::make_shared<ThoughtActionParser>();
                auto modelResponse = std::string("Let's look at the files in the current directory.\
```\
ls -l\
```");
                auto [thought, action] = parser->parse(object{
                    object::pair{std::string("message"), modelResponse}
                }, array<any>());
                expect(thought)->toBe(std::string("Let's look at the files in the current directory.\
"));
                expect(action)->toBe(std::string("ls -l\
"));
            }
            );
            it(std::string("should throw FormatError when no code block"), [=]() mutable
            {
                shared parser = std::make_shared<ThoughtActionParser>();
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), std::string("No code block")}
                    }, array<any>());
                }
                )->toThrow(FormatError);
            }
            );
        }
        );
        describe(std::string("XMLThoughtActionParser"), [=]() mutable
        {
            it(std::string("should parse XML command tags"), [=]() mutable
            {
                auto parser = std::make_shared<XMLThoughtActionParser>();
                auto modelResponse = std::string("Let's look at the files in the current directory.\
<command>\
ls -l\
</command>");
                auto [thought, action] = parser->parse(object{
                    object::pair{std::string("message"), modelResponse}
                }, array<any>());
                expect(thought)->toBe(std::string("Let's look at the files in the current directory."));
                expect(action)->toBe(std::string("ls -l"));
            }
            );
            it(std::string("should throw FormatError when no command tags"), [=]() mutable
            {
                shared parser = std::make_shared<XMLThoughtActionParser>();
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), std::string("No command tags")}
                    }, array<any>());
                }
                )->toThrow(FormatError);
            }
            );
        }
        );
        describe(std::string("EditFormat"), [=]() mutable
        {
            it(std::string("should parse edit format with code blocks"), [=]() mutable
            {
                auto parser = std::make_shared<EditFormat>();
                auto modelResponse = std::string("Let's replace the contents.\
```\
import os\
os.listdir()\
```");
                auto [thought, action] = parser->parse(object{
                    object::pair{std::string("message"), modelResponse}
                }, array<any>());
                expect(thought)->toBe(std::string("Let's replace the contents.\
"));
                expect(action)->toBe(std::string("import os\
os.listdir()\
"));
            }
            );
            it(std::string("should throw FormatError when no code block"), [=]() mutable
            {
                shared parser = std::make_shared<EditFormat>();
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), std::string("No code block")}
                    }, array<any>());
                }
                )->toThrow(FormatError);
            }
            );
        }
        );
        describe(std::string("Identity"), [=]() mutable
        {
            it(std::string("should return input as is"), [=]() mutable
            {
                auto parser = std::make_shared<Identity>();
                auto modelResponse = std::string("Return as is");
                auto [thought, action] = parser->parse(object{
                    object::pair{std::string("message"), modelResponse}
                }, array<any>());
                expect(thought)->toBe(modelResponse);
                expect(action)->toBe(modelResponse);
            }
            );
        }
        );
        describe(std::string("JsonParser"), [=]() mutable
        {
            it(std::string("should parse valid JSON"), [=]() mutable
            {
                auto parser = std::make_shared<JsonParser>();
                auto modelResponse = std::string("{"thought": "List files", "command": {"name": "ls", "arguments": {"path": "."}}}");
                auto [thought, action] = parser->parse(object{
                    object::pair{std::string("message"), modelResponse}
                }, array<any>());
                expect(thought)->toBe(std::string("List files"));
                expect(action)->toBe(std::string("ls ."));
            }
            );
            it(std::string("should throw FormatError for invalid JSON"), [=]() mutable
            {
                shared parser = std::make_shared<JsonParser>();
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), std::string("Not a JSON")}
                    }, array<any>());
                }
                )->toThrow(FormatError);
            }
            );
            it(std::string("should throw FormatError for missing keys"), [=]() mutable
            {
                shared parser = std::make_shared<JsonParser>();
                shared missingKeys = std::string("{"thought": "Missing command key"}");
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), missingKeys}
                    }, array<any>());
                }
                )->toThrow(FormatError);
            }
            );
        }
        );
        describe(std::string("FunctionCallingParser"), [=]() mutable
        {
            it(std::string("should parse function calls successfully"), [=]() mutable
            {
                auto parser = std::make_shared<FunctionCallingParser>();
                auto command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                auto modelResponse = object{
                    object::pair{std::string("message"), std::string("Let's list the files")}, 
                    object::pair{std::string("tool_calls"), array<object>{ object{
                        object::pair{std::string("function"), object{
                            object::pair{std::string("name"), std::string("ls")}, 
                            object::pair{std::string("arguments"), std::string("{}")}
                        }}
                    } }}
                };
                auto [thought, action] = parser->parse(modelResponse, array<std::shared_ptr<Command>>{ command });
                expect(thought)->toBe(std::string("Let's list the files"));
                expect(action)->toBe(std::string("ls"));
            }
            );
            it(std::string("should throw error for missing tool_calls"), [=]() mutable
            {
                shared parser = std::make_shared<FunctionCallingParser>();
                shared command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                expect([=]() mutable
                {
                    parser->parse(object{
                        object::pair{std::string("message"), std::string("No tool calls")}
                    }, array<std::shared_ptr<Command>>{ command });
                }
                )->toThrow(FormatError);
            }
            );
            it(std::string("should throw error for multiple tool calls"), [=]() mutable
            {
                shared parser = std::make_shared<FunctionCallingParser>();
                shared command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                shared multipleCalls = object{
                    object::pair{std::string("message"), std::string("Multiple calls")}, 
                    object::pair{std::string("tool_calls"), array<object>{ object{
                        object::pair{std::string("function"), object{
                            object::pair{std::string("name"), std::string("ls")}, 
                            object::pair{std::string("arguments"), std::string("{}")}
                        }}
                    }, object{
                        object::pair{std::string("function"), object{
                            object::pair{std::string("name"), std::string("cd")}, 
                            object::pair{std::string("arguments"), std::string("{}")}
                        }}
                    } }}
                };
                expect([=]() mutable
                {
                    parser->parse(multipleCalls, array<std::shared_ptr<Command>>{ command });
                }
                )->toThrow(FormatError);
            }
            );
            it(std::string("should throw error for invalid command"), [=]() mutable
            {
                shared parser = std::make_shared<FunctionCallingParser>();
                shared command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                shared invalidCommand = object{
                    object::pair{std::string("message"), std::string("Invalid command")}, 
                    object::pair{std::string("tool_calls"), array<object>{ object{
                        object::pair{std::string("function"), object{
                            object::pair{std::string("name"), std::string("invalid")}, 
                            object::pair{std::string("arguments"), std::string("{}")}
                        }}
                    } }}
                };
                expect([=]() mutable
                {
                    parser->parse(invalidCommand, array<std::shared_ptr<Command>>{ command });
                }
                )->toThrow(FormatError);
            }
            );
            it(std::string("should throw error for invalid JSON arguments"), [=]() mutable
            {
                shared parser = std::make_shared<FunctionCallingParser>();
                shared command = object{
                    object::pair{std::string("name"), std::string("ls")}
                };
                shared invalidJson = object{
                    object::pair{std::string("message"), std::string("Invalid JSON")}, 
                    object::pair{std::string("tool_calls"), array<object>{ object{
                        object::pair{std::string("function"), object{
                            object::pair{std::string("name"), std::string("ls")}, 
                            object::pair{std::string("arguments"), std::string("invalid json")}
                        }}
                    } }}
                };
                expect([=]() mutable
                {
                    parser->parse(invalidJson, array<std::shared_ptr<Command>>{ command });
                }
                )->toThrow(FormatError);
            }
            );
        }
        );
        describe(std::string("FunctionCallingParser error messages"), [=]() mutable
        {
            it(std::string("should format error message correctly"), [=]() mutable
            {
                auto parser = std::make_shared<FunctionCallingParser>();
                auto error = std::make_shared<FunctionCallingFormatError>(std::string("test"), std::string("missing"));
                auto message = parser->formatErrorMessage(error);
                expect(message)->toContain(std::string("did not use any tool calls"));
            }
            );
        }
        );
    }
    );
}

MAIN
